resources/type-system/inference.md

```diff
@@ -970,167 +970,244 @@ with respect to `L` under constraints `C0`
+if (x is num) { x = f(); }`, the variable `x` is promoted to the type `num`
+within the body of the `if` statement. Accordingly, `f()` is type inferred in
+the context `num`, in an effort to reduce the likelihood that the assignment `x
+= f()` will cause the type promotion to be lost. However, if the static type of
+`f()` doesn't wind up being a subtype of `num`, there is no compile-time error;
+`x` is simply demoted back to `Object?` as a side effect of the assignment._
+
+## Static types and soundness
+
+Every elaborated expression has an associated _static type_, which is produced
+as a by-product of the expression inference process. These static types are
+specified in this document using language such as: "inference of `e` produces an
+elaborated expression `m` with static type `T`". _The specification of static
+types in this document should be understood to supersede the specification of
+static types in the existing Dart language specification._
+
+_Informally, we may sometimes speak of the static type of an unelaborated
+expression `e`; when we do, what is meant by this is the static type of the
+elaborated expression `m` that results from performing expression inference on
+`e`. Note in particular that a given expression might have a different static
+type at different points in the code, since the behavior of expression inference
+depends on the context `K`, as well as the flow analysis state._
+
+An invariant of expression inference, known as _soundness_, is that when an
+elaborated expression is executed, it is guaranteed either to diverge, throw an
+exception, or evaluate to a value that is an _instance satisfying_ its static
+type. _Instance satisfying_ is defined as follows: a value `v` is an instance
+satisfying type `T` iff the runtime type of `v` is a subtype of the _extension
+type erasure_ of `T`. _So, for example, every value is considered an instance
+satisfying type `dynamic`, and all values except `null` are considered an
+instance satisfying type `Object`._
+
+_This stands in contrast to the notion of "instance of"; a value `v` is an
+instance of a type `T` only if the runtime type of `v` is __precisely__ `T`._
+
+_The type inference rules below include informal sketches of a proof that
+soundness holds for each expression type. These are non-normative, so they are
+typeset in italics._
+
+## New operations allowed in elaborated expressions
+
+The elaboration process sometimes introduces new operations that are not easily
+expressible using the syntax of Dart. To allow these operations to be specified
+succintly, the syntax of Dart is extended to allow the following forms:
+
+- `@AWAIT_WITH_TYPE_CHECK<T>(m_1)`, represents the following operation:
+
+  - Evaluate `m_1`, and let `v` denote the result.
+
+  - If `v` is an instance satisfying type `Future<T>`, then
+    `@AWAIT_WITH_TYPE_CHECK<T>(m_1)` evaluates to `v`.
+
+  - Otherwise, let `u` be a future satisfying type `Future<T>` that will
```

Comment:
That's a fair critique.

There's an implicit assumption here that I'm only ever going to use `@AWAIT_WITH_TYPE_CHECK<T>(m_1)` in circumstances where I can guarantee that if `T_1` is the static type of `m_1`, then `T` is `flatten(T_1)`. I have a rather complex proof in the "Await expressions" section that since `T` is `flatten(T_1)`, if `v` is not an instance satisfying `Future<T>`, it must be an instance satisfying `T`. And thus, a sound `Future<T>` is guaranteed to exist that will complete to the value `v` at some later point.

As Lasse and I discussed in https://github.com/dart-lang/language/pull/3850#discussion_r1635584296, if we prove the lemma that `T <: FutureOr<flatten(T)>` for all types `T`, then this will all get easier. In particular, what will get easier here is that substituting `T_1` for `T`, we'll have `T_1 <: FutureOr<flatten(T_1)>`. With `T = flatten(T_1)`, that means that `T_1 <: FutureOr<T>`. Since the static type of `m_1` is `T_1`, it follows that `v` is an instance satisfying `T_1`, so by transitivity it must be an instance satisfying `FutureOr<T>`. That in turn means it must either be an instance satisfying `Future<T>` or an instance satisfying `T`. Since the "if" bullet above handles the case where it's an instance satisfying `Future<T>`, in this "otherwise" bullet, we know that `v` is an instance satisfying `T`.

For now I've left a short TODO comment sketching out this logic; I'll fill it out in the follow-up PR.

Reply:
This should be addressed by https://github.com/dart-lang/language/pull/3920.